protocols/request-response/src/cbor.rs

```diff
@@ -0,0 +1,244 @@
+    Resp: Send + Clone + Serialize + DeserializeOwned + 'static,
+{
+    inner: crate::Behaviour<codec::Codec<Req, Resp>>,
+}
+
+impl<Req, Resp> Behaviour<Req, Resp>
+where
+    Req: Send + Clone + Serialize + DeserializeOwned,
+    Resp: Send + Clone + Serialize + DeserializeOwned,
+{
+    pub fn new<I>(protocols: I, cfg: Config) -> Self
+    where
+        I: IntoIterator<
+            Item = (
+                <codec::Codec<Req, Resp> as crate::Codec>::Protocol,
+                ProtocolSupport,
+            ),
+        >,
+    {
+        Behaviour {
+            inner: crate::Behaviour::new(codec::Codec::default(), protocols, cfg),
+        }
+    }
+
+    pub fn send_request(&mut self, peer: &PeerId, request: Req) -> RequestId {
+        self.inner.send_request(peer, request)
+    }
+
+    pub fn send_response(&mut self, ch: ResponseChannel<Resp>, rs: Resp) -> Result<(), Resp> {
+        self.inner.send_response(ch, rs)
+    }
+
+    pub fn add_address(&mut self, peer: &PeerId, address: Multiaddr) {
+        self.inner.add_address(peer, address)
+    }
+
+    pub fn remove_address(&mut self, peer: &PeerId, address: &Multiaddr) {
+        self.inner.remove_address(peer, address)
+    }
+
+    pub fn is_connected(&self, peer: &PeerId) -> bool {
+        self.inner.is_connected(peer)
+    }
+
+    pub fn is_pending_outbound(&self, peer: &PeerId, request_id: &RequestId) -> bool {
+        self.inner.is_pending_outbound(peer, request_id)
+    }
+
+    pub fn is_pending_inbound(&self, peer: &PeerId, request_id: &RequestId) -> bool {
+        self.inner.is_pending_inbound(peer, request_id)
+    }
+}
+
+mod codec {
+    use async_trait::async_trait;
+    use futures::prelude::*;
+    use futures::{AsyncRead, AsyncWrite};
+    use libp2p_core::upgrade::{read_length_prefixed, write_length_prefixed};
+    use libp2p_swarm::StreamProtocol;
+    use serde::{de::DeserializeOwned, Serialize};
+    use std::{io, marker::PhantomData};
+
+    /// Max request size in bytes
+    const REQUEST_SIZE_MAXIMUM: usize = 1024 * 1024;
+    /// Max response size in bytes
+    const RESPONSE_SIZE_MAXIMUM: usize = 10 * 1024 * 1024;
+
+    #[derive(Debug, Clone)]
+    pub struct Codec<Req, Resp> {
+        phantom: PhantomData<(Req, Resp)>,
+    }
+
+    impl<Req, Resp> Default for Codec<Req, Resp> {
+        fn default() -> Self {
+            Codec {
+                phantom: PhantomData,
+            }
+        }
+    }
+
+    #[async_trait]
+    impl<Req, Resp> crate::Codec for Codec<Req, Resp>
+    where
+        Req: Send + Clone + Serialize + DeserializeOwned,
+        Resp: Send + Clone + Serialize + DeserializeOwned,
+    {
+        type Protocol = StreamProtocol;
+        type Request = Req;
+        type Response = Resp;
+
+        async fn read_request<T>(&mut self, _: &Self::Protocol, io: &mut T) -> io::Result<Req>
+        where
+            T: AsyncRead + Unpin + Send,
+        {
+            let vec = read_length_prefixed(io, REQUEST_SIZE_MAXIMUM).await?;
```

Comment:
Here, can be we try something like:

```suggestion
            let mut vec = Vec::new();
            
            loop {
            	if vec.len() > REQUEST_SIZE_MAXIMUM {
            		return Err(...) // Something with max size exceeded
            	}
            	
            	let num_bytes = (&mut io).take(1_000).read_to_end(&mut vec).await?;
            	
            	if num_bytes == 1000 {
            		break;
            	}
            }
```